### ISSUE TITLE
Director aiming

### ISSUE THREAD

**Issue by jamesol1** [TylerS1066/Movecraft-Cannons#2]
Hi Tyler,
I was wondering if you could add a Director for cannon projectiles. I spoke to the Cannons owner about this but hes not active on the project anymore. Also i think it would involve a lot more changes to pull off. 

Cannons allows one to auto reload and using redstone fire off a selection of them in one go. If one could direct the aiming this could allow for a wider variety of user with cannons and movecraft! I mean you could have ships with 3 large cannons that could aim similar to tnt turrets.

**Comment by TylerS1066** [comment]
I believe the Cannons plugin has a built in method to adjust the aim of cannons individually & permanently, but adding an AA-Director style system would be an interesting addition to this.

I currently am focusing my time on working on [Movecraft-Warfare](https://github.com/APDevTeam/Movecraft-Warfare), and as such I won't have the bandwidth to implement this currently.  If somebody wants to implement this, they could follow the template of directors done in Movecraft-Combat and instead use the Cannons API events.

**Comment by jamesol1** [comment]
Unfortunetly you cant aim more then one cannon at once with the cannons plugin. I did suggest it to Pavlov but hes not working on the plugin anymore atleast outside of bug fixes. Ah no worries. If someone could add that would be great!

**Comment by DerPavlov** [comment]
I added some features for Cannons to aim, fire and reload with multiple cannons simultaniously. What would you need to integrated to Movecraft?
There was a discussion to group cannons which are on a ship when the ship is created.
Maybe limiting linking Cannons to ships only to give them devastating firepower.

**Comment by TylerS1066** [comment]
I'm currently pre-occupied with preparing the other addons and my servers for an update to 1.14.4, so this addon has fallen to the bottom of my priority list.

Once I do bring my attention back to this, my first priority will be #9 and #10, along with the two protection integration problems (https://github.com/TylerS1066/Movecraft-Towny/issues/3 and https://github.com/APDevTeam/Movecraft-WorldGuard/issues/4).  Once all of those are fixed, I'd like to implement #1.  Once all that is completed, the plugin will be a stable base we can start looking into adding features like directors etc.  I am however fully accepting of PRs, so long as they don't conflict with the goals of the plugin.

All that said, if it is possible to hold references to `Cannon` objects for the entire duration of the pilot period of a craft, it may actually simplify my implementation of cannon movement to detect cannons on piloting, and keep track of them.  This would also make it easier to implement grouped cannon features in the future.  Is this a reasonable way to try and track/move cannons (aka, will they continue to be valid references after being moved and rotated)?

**Comment by jamesol1** [comment]
Hi Tyler, just wondering if the Cannon plugin director aiming was added or looked at in the end. I would quite like this addition, not sure how doable it is, noticed Pavlov didn't reply to your last question.

**Comment by DaarkKarma** [comment]
Currently the default cannons plugin has a system in place where (if set in the cannons file) you can link multiple cannons of the same type and aim them with a clock. This however is seperate from movecraft and doesn't really take movecraft in mind. Linking doesn't focus on ships meaning any cannon you own as long as your close to it will link even if it isnt on the ship. 
It is however a decent middle ground to use until a director or similar system is added to the plugin.

**Comment by jamesol1** [comment]
Thanks for mentioning that to me Karma. I did test out the linked cannons feature added, as you mentioned it has its quirks. I do personally prefer the Cannon director means of aiming. Being able to use that without players needing quite advanced and large redstone vanilla cannons wold be great. If anyone is able to implement it that would be awesome of course. I know Pavlov did also allow primed_tnt to be shot from his cannons, although this doesn't seem to be influenced by Movecrafts Cannon directors when i last tested it.

**Comment by DaarkKarma** [comment]
primed tnt causes issues in the cannons plugin as well, apparently it's buggy and not the best option to use.
I do agree though that there should be something in place. I know some people's private forks of movecraft have a middle man kind of method, a way similar to linked cannons but it's ship specific. It works like the cannon director but uses the clock instead of a sign that needs to be activated. 
Perhaps something like that would be useful

**Comment by jamesol1** [comment]
True, primed tnt did perform very buggy aiming wise. Also you'll notice linked cannon aiming adjusts the cannons the same amount... meaning they wont hone in on one location like they do with Cannon directors. So if they are somewhat spread out you'll have 3 spaced out projectiles shooting out. You also cant move very far from the primary cannon when aiming in clock mode. It's a shame there isnt any other block cannon plugins out there. That does sound like a good solution.

**Comment by TylerS1066** [comment]
Unfortunately the status is still similar to what it was when I made that comment, but all the prerequisite issues have been resolved.

Since that comment, the scope of updating the various addons has drastically increased, including ground up rewrites of Movecraft-Combat and Movecraft-Repair, with heavy updates to Movecraft-Warfare.  On top of that, I've realized that Movecraft itself needs a lot more work and maintenance than I once thought.

Unfortunately, since this plugin isn't used on the Airship Pirates servers, it's not going to be on my priority list, and is in fact lower than https://github.com/APDevTeam/Movecraft-Towny/issues/4 since that plugin no longer functions on the latest Towny versions.  That said, I am 100% open to another developer coming in and implementing this feature.

**Comment by jamesol1** [comment]
No worries Tyler that makes sense. I'm **open** to the idea of comissioning someone to do it, although i'm not sure how much that sort of thing would be expected to cost. By all means someone reach out if you are familiar with Cannons and Movecraft functionality.
